pyanaconda/modules/payload/base/source_base.py

```diff
@@ -0,0 +1,111 @@
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+from abc import ABCMeta, abstractmethod
+
+from pyanaconda.core.signal import Signal
+from pyanaconda.dbus.publishable import Publishable
+from pyanaconda.modules.common.base import KickstartBaseModule
+
+
+class PayloadSourceBase(KickstartBaseModule, Publishable, metaclass=ABCMeta):
+    """Base class for all the payload source modules.
+
+    This object contains all API shared for all the sources. Everything in this object have
+    to be implemented by a source to be used.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._is_ready = False
+        self.is_ready_changed = Signal()
+
+    @property
+    @abstractmethod
+    def kind(self):
+        """Get type of this source object.
+
+        :return: type of this source
+        :type: value of payload.base.constants.SourceType
+        """
+        pass
+
+    @property
+    def is_ready(self):
+        """This source is ready for the installation to start.
+
+        This method will not be part of the public API. There is no need for others than the
+        payload owner to see the status of the source.
+
+        :rtype: bool
+        """
+        return self._is_ready
+
+    def _set_is_ready(self, value):
+        """Set if the source is ready to be used for the installation process.
+
+        :param value: True if the source is ready
+        :type value: bool
+        """
+        self._is_ready = value
+        self.is_ready_changed.emit()
+
+    @abstractmethod
+    def set_up_with_tasks(self):
+        """Prepare this payload source.
+
+        Do everything it requires to be able to use this source for installation.
+
+        Call self._set_is_ready(True) method when this task is finished successfully.
+
+        This method will not be part of the public API. Having this private only will prevent race
+        conditions and there is no real use outside of payload.
+
+        :return: list of tasks to prepare this source
+        :rtype: list[task]
+        """
+        pass
+
+    @abstractmethod
+    def tear_down_with_tasks(self):
+        """Tear down this payload source.
+
+        Cleanup everything done by the setup method.
+
+        Call self._set_is_ready(False) method when this task is finished successfully.
+
+        This method will not be part of the public API. Having this private only will prevent race
+        conditions and there is no real use outside of payload.
+
+        :return: list of tasks to tear down this source
+        :rtype: list[task]
+        """
+        pass
+
+    @abstractmethod
+    def validate(self):
+        """Do a quick validation of the given source.
```

Comment:
"of the given source" -> "of the source" ?

Reply:
Fixed.